### ISSUE TITLE
DType of update depends on whether bfloat16 learning rate is passed as float, jax scalar or numpy scalar

### ISSUE THREAD

**Issue by colehaus** [google-deepmind/optax#1405]
```py
model = jnp.ones(8, bfloat16)
tx = optax.adam(learning_rate=1e-3)
opt_state = tx.init(model)
jax.jit(tx.update)(model, opt_state, model)[0].dtype
```
results in `dtype(bfloat16)`.

```py
model = jnp.ones(8, bfloat16)
tx = optax.adam(learning_rate=jnp.array(1e-3, bfloat16))
opt_state = tx.init(model)
jax.jit(tx.update)(model, opt_state, model)[0].dtype
```
also results in `dtype(bfloat16)`.

```py
model = jnp.ones(8, bfloat16)
tx = optax.adam(learning_rate=np.array(1e-3, bfloat16))
opt_state = tx.init(model)
jax.jit(tx.update)(model, opt_state, model)[0].dtype
```

results in `dtype('float32')`.

I believe this behavior is arising because of different type promotion rules between numpy and jax in functions like this:

https://github.com/google-deepmind/optax/blob/fabb0f8ac477245d1f8b63ab435bf085789abd40/optax/tree_utils/_tree_math.py#L389

`(np.ones(8, bfloat16) ** 0.5).dtype` is `dtype('float32')` while `(jnp.ones(8, bfloat16) ** 0.5).dtype` is `dtype(bfloat16)`.

I realize JAX is on record as having different type promotion rules (https://docs.jax.dev/en/latest/type_promotion.html) but the end result in optax seems pretty unexpected and undesirable in cases like this.

**Comment by rdyro** [comment]
This is unfortunately the result of numpy dtype promotion, consider:
```python
import numpy as np, ml_dtypes
(1 * np.ones((), "bfloat16")).dtype
# prints "dtype('float64')"
```
which in optax happens when we scale the learning_rate by -1 to perform gradient descent `-1 * learning_rate`.

I'm not sure we can do anything apart from warning about this in the docs. This is definitely not obvious behavior, so if you have any suggestions for addressing this, we'd be happy to brainstorm.

**Comment by rdyro** [comment]
Closing as this behavior is confusing, but expected.
